File: Unit 2-Data Structure/Queue/As Array/Queue.java
class AQueue<T>
{
    private static final int MAX_SIZE = 100;
    private T queue[];
    private int front, rear;

    @SuppressWarnings("unchecked")
    public AQueue() {
        queue = (T[]) new Object[MAX_SIZE];
        front = rear = -1;
    }

    public void enqueue(T value) {
        if (isFull()) {
            System.out.println("Queue Overflow");
        } else {
            if (isEmpty()) {
                front = rear = 0;
            } else {
                rear = (rear + 1) % MAX_SIZE;
            }
            queue[rear] = value;
        }
    }

    public T dequeue() {
        if (isEmpty()) {
            System.out.println("Queue Underflow");
            return null;
        } else {
            T item = queue[front];
            if (front == rear) {
                front = rear = -1;
            } else {
                front = (front + 1) % MAX_SIZE;
            }
            return item;
        }
    }

    public boolean isEmpty() {
        return front == -1 && rear == -1;
    }

    public boolean isFull() {
        return (rear + 1) % MAX_SIZE == front;
    }
}


class Queue
{
    public static void main(String[] args)
    {
            AQueue<Integer> myQueue = new AQueue<>();

            for (int i = 1; i < 10; i++)
            {
                myQueue.enqueue(i);
            }

            System.out.println("Dequeue: " + myQueue.dequeue());
            System.out.println("Is the queue empty? " + (myQueue.isEmpty() ? "Yes" : "No"));
        }
    }

File: Unit 2-Data Structure/Queue/As LinkedList/Queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LQueue:
    def __init__(self):
        self.front = self.rear = None

    def enqueue(self, value):
        new_node = Node(value)
        if self.is_empty():
            self.front = self.rear = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node

    def dequeue(self):
        if self.is_empty():
            print("Queue Underflow")
            return None
        else:
            item = self.front.data
            self.front = self.front.next
            if self.front is None:
                self.rear = None
            return item

    def is_empty(self):
        return self.front is None and self.rear is None


if __name__ == "__main__":
    my_queue = LQueue()

    for i in range(1, 10):
        my_queue.enqueue(i)

    print("Dequeue:", my_queue.dequeue())
    print("Is the queue empty?", "Yes" if my_queue.is_empty() else "No")
